hal/spi_async_api.h

```diff
@@ -0,0 +1,178 @@
+/** \ingroup hal_spi */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_SPI_API_ASYNC_H
+#define MBED_SPI_API_ASYNC_H
+
+#include "hal/spi_api.h"
+
+#if DEVICE_SPI
+#if DEVICE_SPI_ASYNCH
+
+/** \defgroup hal_spi_async SPI: Asynchronous API.
+ * Asynchronous extension to the SPI HAL API.
+ *
+ * # Programming model
+ * This is an OOP like API. spi_async_transfer_t is used a "this".<br/>
+ * Methods not having this type as their first parameter are **class** methods.<br/>
+ * Methods with an spi_t object as their first paremeter are **instance** methods.<br/>
+ * `spi_async_transfer_new()` acts as the "constructor" of this *class*.<br/>
+ * `spi_async_transfer_free()` acts as the "desctructor" of this *class*.
+ * 
+ * This API extends the hal_spi API.
+ * 
+ * # Defined behaviours
+
+## Defined behaviours
+ * - A repeating transfer must not restart until the invokation of the completion callback has not returned.
+ * - The implementation may use interrupt or 
+ * 
+ * - spi_async_transfer_new(..)
+ *   - enqueues a transfer request that will processed as the device become available.
+ *   - on successfull enqueueing returns a counted reference to the related hal implementation
+ *     resource.
+ *     @warning if `repeat` is `false` then any element passed to this function must live at least
+ *     until the callback is invoked.
+ *     @warning if `repeat` is `true` then any element passed to this function must live at least
+ *     until the transfer is freed.
+ * - spi_async_transfer_free(..)
+ *   - aborts/cancels a transaction if not completed yet and release the associated memory.
+ *   - it does **NOT** free the spi_transfer_args_t **nor** the referenced buffers.
+ * - spi_event_f(..)
+ *   - may update the spi_transfer_args_t structure initialy given to `spi_async_transfer_new(..)`
+ *     on a repeating transfer.
+ *   - does not block/busy wait/perform heavy calculation.
+ *   - does not perfom any rtos unsafe operation.
+ *
+ * # Undefined behaviours
+ * - Calling any method on a `spi_async_tranfer_t` after calling `spi_async_free_handle()`.
+ * - Passing invalid pointers to any method.
+ * - Freeing buffers before freeing the associated transaction.
+ *
+ * # Lexicon
+ * ## `Completion`/`Cancelletion`/`Abortion`
```

Comment:
Consider a different word/phrase in place of "abortion" (here and elsewhere), to avoid unrelated connotations. "Abort" or "aborting" are fine if the context fits, IMO.